### PR TITLE
[TE/JAX] Prototype for New XLA Custom Calls with FFI

### PR DESCRIPTION
**Pull request by phu0ngng**
# Description

This PR introduced a prototype for converting TE/JAX custom calls to the new XLA Custom Call with FFI.
These primitives are implemented with the new custom calls:
- `CastTranspose`
- `ActLu`
- `DActLu`

`FFI Datatype` with FP8 types has been added to XLA in this PR (openxla/xla#13856). 
All related tests passed.

 Verified: custom calls are captured in CudaGraph by default.

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [ ] Code refractor

# Checklist:

- [x] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [ ] The functionality is complete
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: transformer_engine/jax/csrc/extensions/activation.cpp [hunk 1]
```diff
@@ -132,9 +160,8 @@ void DActLu(cudaStream_t stream, void **buffers, const char *opaque, size_t opaq
 
   const auto &desc = *UnpackOpaque<CustomCallCommonDescriptor>(opaque, opaque_len);
   auto m = desc.shape.dims[0];
-  auto n = desc.shape.dims[1];
+  auto n = desc.shape.dims[2];
```
Comment (phu0ngng) [inline transformer_engine/jax/csrc/extensions/activation.cpp:163]: This is supposed to be wrong and the unit test should have failed.

File: transformer_engine/jax/cpp_extensions/custom_call.py [hunk 2]
```diff
@@ -18,7 +19,15 @@
     pass
 
 for _name, _value in transformer_engine_jax.registrations().items():
-    xla_client.register_custom_call_target(_name, _value, platform="CUDA")
+    if "_ffi" in _name:
+        if jax_version_meet_requirement():
+            # TODO: add prep, init, exec
+            # traits = 1 (i.e. COMMAND_BUFFER_COMPATIBLE) enabled cudaGraph
+            xla_client.register_custom_call_target(
+                _name, _value, platform="CUDA", api_version=1, traits=1
```
Comment (zlsh80826) [inline transformer_engine/jax/cpp_extensions/custom_call.py:27]: Would be great to add a IntEnum for `api_version` or at least a comment to describe it.
```
class CustomCallAPIVersion(IntEnum):
    OPAQUE = 0
    FFI = 1
```
```
xla_client.register_custom_call_target(
   _name, _value, platform="CUDA",
   api_version=CustomCallAPIVersion.FFI.value)
```

File: transformer_engine/jax/cpp_extensions/custom_call.py [hunk 3]
```diff
@@ -93,6 +102,40 @@ def custom_caller(name, args, opaque, has_side_effect, **kwargs):
...
+def custom_caller_with_ffi(name, args, backend_config, **kwargs):
+    """
+    New XLA custom call warpper
+    """
+    return mlir.custom_call(
+        call_target_name=name,
+        result_types=args.output_types,
+        operands=args.operands,
+        backend_config=backend_config,
+        api_version=4,
```
Comment (zlsh80826) [inline transformer_engine/jax/cpp_extensions/custom_call.py:139]: What does 4 mean in this context?

File: transformer_engine/jax/cpp_extensions/misc.py [hunk 4]
```diff
@@ -128,3 +131,12 @@ def multidim_transpose(shape, static_axis_boundary, transpose_axis_boundary):
         *shape[transpose_axis_boundary:],
         *shape[transpose_start_idx:transpose_axis_boundary],
     )
+
+
+def jax_version_meet_requirement(version: str = "0.4.29"):
```
[the remark this answers is not in the document]
Comment (phu0ngng) [reply transformer_engine/jax/cpp_extensions/misc.py:136]: Hi, the function could be reused in the future if there are any cases where we have some differences between the two Jax versions, but `is_ffi_available()` is fine too.

File: transformer_engine/jax/cpp_extensions/custom_call.py [hunk 5]
```diff
@@ -40,9 +49,9 @@ def __init__(
         self.operand_layouts = CustomCallArgsWrapper.generate_layouts(
             operand_shapes, operand_specific_layouts
         )
-        output_shapes = [x.shape for x in output_types]
+        self.output_shapes = [x.shape for x in output_types]
```
[the remark this answers is not in the document]
Comment (phu0ngng) [reply transformer_engine/jax/cpp_extensions/custom_call.py:52]: This was a mistake indeed.

File: transformer_engine/jax/cpp_extensions/custom_call.py (hunk 3 above)
Comment (phu0ngng) [reply transformer_engine/jax/cpp_extensions/custom_call.py:139]: `api_version=2` for the old custom call and `4` for the new one ([example](https://github.com/google/jax/blob/988ff27dfa03bb2102efefb48885ffe582be834a/jaxlib/gpu_prng.py#L123)).
I first implemented this`custom_caller_with_ffi()` when the `ffi.ffi_lowering()` was not yet there. 
Since the `ffi.ffi_lowering` seems to work fine for our needs now, I'm considering to remove this function.

File: transformer_engine/jax/cpp_extensions/custom_call.py [hunk 6]
```diff
@@ -17,8 +19,25 @@
...
+class CustomCallAPIVersion(IntEnum):
+    """Enum for selecting between old and new custom call registration API"""
+
+    OPAQUE = 0
+    FFI = 1
+
+
 for _name, _value in transformer_engine_jax.registrations().items():
-    xla_client.register_custom_call_target(_name, _value, platform="CUDA")
+    if "_ffi" in _name:
```
Comment (nouiz) [inline transformer_engine/jax/cpp_extensions/custom_call.py:31]: Check that it is at the end? `"ddd".endswith("d")`

File: tests/jax/test_custom_call_compute.py [hunk 7]
```diff
@@ -432,7 +432,7 @@ def ref_act_lu(inputs):
     def primitive_func(self, inputs):
         return jnp.mean(activation_lu(inputs, activation_type=self.activation_type))
 
-    @pytest.mark.parametrize("shape", [(32, 1, 64), (64, 1, 256)])
+    @pytest.mark.parametrize("shape", [(32, 1, 64), (16, 64, 1, 256)])
```
Comment (nouiz) [inline tests/jax/test_custom_call_compute.py:435]: Why the change in shape?

Comment (phu0ngng) [reply tests/jax/test_custom_call_compute.py:435]: Hi, I found out that we only had three-dimensional input cases, so I changed it to have a case with more than 3 dimensions.

File: transformer_engine/jax/csrc/extensions/activation.cpp [hunk 8]
```diff
@@ -182,6 +208,76 @@ void DActLu(cudaStream_t stream, void **buffers, const char *opaque, size_t opaq
...
+
+  auto input_shape = std::vector<size_t>{m, n};
+  auto act_input_shape = std::vector<size_t>{m, n * act_len};
+  auto output_shape = std::vector<size_t>{m, n * act_len};
+
+  auto input_tensor = TensorWrapper(input, input_shape, static_cast<DType>(in_dtype));
+  auto act_input_tensor = TensorWrapper(act_input, act_input_shape, static_cast<DType>(in_dtype));
+  auto output_tensor = TensorWrapper(output, output_shape, static_cast<DType>(out_dtype));
+
+  auto act_type = static_cast<NVTE_Activation_Type>(act_enum);
```
Comment (denera) [inline transformer_engine/jax/csrc/extensions/activation.cpp:234]: This looks like duplicate code with `DActLu()`. Is there a reason why we can't put it into a `DActLuImpl()` function and then call it from both `DActLu()` and `DActLuFFI()`?

**Review by denera — CHANGES_REQUESTED** [review]

File: transformer_engine/jax/csrc/extensions/activation.cpp (hunk 8 above)
Comment (phu0ngng) [reply transformer_engine/jax/csrc/extensions/activation.cpp:234]: I did notice it.
But `DActLu()` will be removed eventually so there won't be this code duplication.

**Review by denera — APPROVED** [review]

File: transformer_engine/jax/cpp_extensions/custom_call.py [hunk 9]
```diff
@@ -17,8 +19,25 @@
...
+    FFI = 1
+
+
 for _name, _value in transformer_engine_jax.registrations().items():
-    xla_client.register_custom_call_target(_name, _value, platform="CUDA")
+    if _name.endswith("_ffi"):
+        if is_ffi_enabled():
+            # COMMAND_BUFFER_COMPATIBLE i.e. cudaGraph enabled by default
+            xla_client.register_custom_call_target(
+                _name, _value, platform="CUDA", api_version=CustomCallAPIVersion.FFI.value
```
Comment (nouiz) [inline transformer_engine/jax/cpp_extensions/custom_call.py:35]: This don't seem to pass the cudaGraph traits.
But the comment and the PR description tell it enable cudaGraph.
Do this need a PR to enable it?
https://github.com/openxla/xla/blob/de99c1bca9ac1ca5faee3c09f46f56b2c0060f89/xla/python/xla_client.py#L589
